File: alura/php_oo_001/src/Account.php
<?php

class Account
{
  private Holder $holder;
  private string $number;
  private float $balance;

  public function __construct(Holder $holder, string $number)
  {
    $this->holder = $holder;
    $this->number = $number;
    $this->balance = 0;
  }

  public function deposit(float $value): void
  {
    $this->balance += $value;
  }

  public function withdraw(float $value): void
  {
    $this->balance -= $value;
  }

  public function balance(): float
  {
    return $this->balance;
  }
}

File: alura/go_oo/004/bank/account/account.go
package account

import "errors"

type Account struct {
	Holder  string
	Agency  string
	Number  string
	Balance int
}

func (self *Account) Withdraw(value int) error {
	if value <= 0 {
		return errors.New("value to withdraw must be positive")
	}

	balance := self.Balance - value

	if balance < 0 {
		return errors.New("value to withdraw is not available")
	}

	self.Balance = balance

	return nil
}

func (self *Account) Deposit(value int) error {
	if value <= 0 {
		return errors.New("value to deposit must be positive")
	}

	self.Balance += value

	return nil
}

func (self *Account) Transfer(value int, target *Account) error {
	err := self.Withdraw(value)

	if err != nil {
		return err
	}

	err = target.Deposit(value)

	if err != nil {
		return err
	}

	return nil
}
